app/src/main/java/my/noveldokusha/scraper/sources/NovelHall.kt

```diff
@@ -0,0 +1,86 @@
+package my.noveldokusha.scraper.sources
+
+import my.noveldokusha.data.BookMetadata
+import my.noveldokusha.data.ChapterMetadata
+import my.noveldokusha.scraper.*
+import org.jsoup.nodes.Document
+
+/**
+ * Novel Example
+ * https://www.novelhall.com/Apollo-s-Heart-15505/
+ * Chapter url example:
+ * https://www.novelhall.com/Apollo-s-Heart-15505/3304460.html
+ */
+class NovelHall : SourceInterface.catalog
+{
+    override val name = "Novel Hall"
+    override val baseUrl = "https://www.novelhall.com/"
+    override val catalogUrl = "https://www.novelhall.com/all.html"
+    override val language = "English"
+
+    override suspend fun getChapterTitle(doc: Document): String? = null
+
+    override suspend fun getChapterText(doc: Document): String
+    {
+        doc.selectFirst("#htmlContent")!!.let {
+            return textExtractor.get(it)
+        }
+    }
+
+    override suspend fun getBookCoverImageUrl(doc: Document): String?
+    {
+        return doc.selectFirst(".book-img.hidden-xs")
+            ?.select("img[src]")
+            ?.attr("src")
+    }
+
+    override suspend fun getBookDescripton(doc: Document): String?
+    {
+        return doc.selectFirst("span.js-close-wrap")
+            ?.let { textExtractor.get(it) }
+    }
+
+    override suspend fun getChapterList(doc: Document): List<ChapterMetadata>
+    {
+        return doc.select("#morelist li.post-1 a").map {
+            ChapterMetadata(title = it.text(), url = it.attr("href"))
+        }.reversed()
+    }
```

Comment:
Tested, doesn't work. Needs a fix.

Reply:
here is python version ```  for a in soup.select('div#morelist.book-catalog ul li a'):
            chap_id = len(self.chapters) + 1
            vol_id = 1 + len(self.chapters) // 100
            if len(self.volumes) < vol_id:
                self.volumes.append({ 'id': vol_id })
            # end if
            self.chapters.append({
                'id': chap_id,
                'volume': vol_id,
                'url':  self.absolute_url(a['href']),
                'title': a.text.strip() or ('Chapter %d' % chap_id),
            })
        # end for```

Reply:
But this PR needs the kotlin version :=)